### ISSUE TITLE
yield & component helper => two way data binding not working ?

### ISSUE THREAD

**Issue by plcarmel** [emberjs/ember.js#13061]
DEBUG: -------------------------------
DEBUG: Ember      : 2.4.1
DEBUG: Ember Data : 2.4.0
DEBUG: jQuery     : 2.2.1
DEBUG: -------------------------------

I don't know if this is a bug but it seems to contradict the documentation at
[http://emberjs.com/api/classes/Ember.Templates.helpers.html#toc_nested-usage]

This is the example that is given:
`{{yield (hash nameInput=(component "my-input-component" value=model.name placeholder="First Name"))}}`

I tried it but without being able to achieve two way data binding.

## To replicate the problem ##

templates/components/test-binding.hbs:
`{{yield (component 'my-input' value=value)}}`

templates/components/my-input.hbs:
`{{input value=value}}`

templates/application.hbs:
```
{{#test-binding value=value as |component|}}
  {{component component}}
{{/test-binding}}

value: {{value}}
```

## A work-around ##

It works it the value is encapsulated inside a hash.  Please note that the way I do it differs from the example in the documentation, which is not working.

templates/components/test-binding-2.hbs:
`{{yield (component 'my-input-2' model=model)}}`

templates/components/my-input-2.hbs:
`{{input value=model.value}}`

templates/application.hbs:
```
{{#test-binding-2 model=model as |component|}}
  {{component component}}
{{/test-binding-2}}

model.value: {{model.value}}
```

controllers/application.js:
```
import Ember from 'ember';
export default Ember.Controller.extend({
  model: { value: 'Hello world' }
});
```

**Comment by GCorbel** [comment]
Can you try simply : 

```
{{#test-binding value=value as |component|}}
  {{component}}
{{/test-binding}}

value: {{value}}
```

I find `{{component component}}` a bit weird.

**Comment by plcarmel** [comment]
From the same documentation:

"Contextual components are created using the nested form of the component helper, and may be passed attrs and positional params. Contextual components **must be invoked with a . in their path, unless they are being passed to the invoking {{component helper.**"

**Comment by pixelhandler** [comment]
@plcarmel can you create an ember-twiddle example that demonstrate the issue? Also I'm curious if 2-way binding is supposed to function, the docs mention binding, but not 2-way.

**Comment by plcarmel** [comment]
In the example given, a (custom) input component is used. So, it is implicit that two way binding should work in that example. There is no action registered on the component, no one-way data binding would render the component useless.

Anyway, things need to be clarified. I will create the Ember Twiddle soon (one week max), stay tuned.

**Comment by courajs** [comment]
I may have found the same bug. I don't think `yield` is a factor here. Here's my reduced test case: [Twiddle](https://ember-twiddle.com/540998d9d06aa72e746f?fileTreeShown=false&numColumns=2&openFiles=x-thing.template.hbs%2Capplication.template.hbs). Let me know if you think this is the same issue.

**Comment by courajs** [comment]
@Serabe You have a lot of knowledge on `(component` right?

**Comment by Serabe** [comment]
Looking into it

**Comment by plcarmel** [comment]
Yes, I think this is the same issue.

**Comment by Serabe** [comment]
**TL;DR** Found the issue. I think I fixed it but a few more tests are needed. I hope to have it done pretty soon.

For normal component invocation, Ember adds a `mut` automatically. Sadly, this does not happen for `contextual components`. I am adding a compiler plugin for doing so, but needs more testing before being published.

If everything goes ok, I hope to have it finished before next week. But again, only if everything goes ok.

**Comment by courajs** [comment]
@Serabe Thanks for the update, and for working on it :)

**Comment by krisselden** [comment]
If you manually use `value=(mut model.name)` does it work?

**Comment by Serabe** [comment]
Yes, it does. The problem is fixed but I need testing it a bit more before
pushing a PR

**Comment by Serabe** [comment]
Fixed in #13177
